File: tests/domain/station/station.test.ts
import { departementCode, stationCode } from "@domain/station/station";

/**
 * Valeurs relevées par appel réel le 2026-08-01 sur
 * `/v2/hydrometrie/referentiel/stations?code_station=K447001001` :
 *   code_station "K447001001" (10) · code_site "K4470010" (8)
 *   code_departement "41" (chaîne) · en_service true
 */
describe("code station (C-05)", () => {
  it("accepte un code à dix caractères", () => {
    expect(stationCode("K447001001")).toBe("K447001001");
  });

  it("refuse un code site à huit caractères", () => {
    // Interroger un code site renvoie chaque mesure EN DOUBLE (C-05).
    // Le refuser au plus tôt évite de propager des doublons jusqu'à la vue.
    expect(() => stationCode("K4470010")).toThrow(/dix caractères/);
  });

  it("nomme le piège dans le message d'erreur", () => {
    // Un message qui dit seulement « longueur invalide » laisserait chercher.
    expect(() => stationCode("K4470010")).toThrow(/double/i);
  });

  it("refuse une chaîne vide plutôt que de la laisser passer", () => {
    expect(() => stationCode("")).toThrow(/dix caractères/);
  });
});

describe("code département", () => {
  it("conserve le zéro de tête", () => {
    // « 01 » traité comme un nombre deviendrait 1, et ne correspondrait plus à
    // aucun département côté API. Le type interdit de le confondre avec un
    // nombre.
    expect(departementCode("01")).toBe("01");
    expect(departementCode("41")).toBe("41");
  });

  it("accepte les codes d'outre-mer à trois caractères", () => {
    // 971 à 976 — et 974 n'a AUCUN point ONDE, ce que BR-007 oblige à dire.
    expect(departementCode("974")).toBe("974");
  });

  it("accepte la Corse", () => {
    expect(departementCode("2a")).toBe("2A");
    expect(departementCode("2B")).toBe("2B");
  });

  it("refuse un format inattendu", () => {
    expect(() => departementCode("4")).toThrow(/département/i);
    expect(() => departementCode("0041")).toThrow(/département/i);
  });
});

describe("forme du code station, mesurée sur le référentiel", () => {
  it("accepte les deux formes réellement observées", () => {
    // Relevé le 2026-08-15 sur les 4 150 stations en service :
    //   3 974 en « A999999999 » — une lettre puis neuf chiffres
    //     176 en « 9999999999 » — dix chiffres, les stations des DOM
    // Restreindre à ^[A-Z]\d{9}$ rejetterait ces 176 stations.
    expect(stationCode("K447001001")).toBe("K447001001");
    expect(stationCode("1011000101")).toBe("1011000101");
  });

  it("refuse dix caractères qui ne sont pas un code", () => {
    // Sans contrôle de forme, ces valeurs partaient en `code_entite` vers
    // Hub'Eau, revenaient vides, et l'écran affichait « pas de donnée » —
    // un symptôme que BR-007 interdit de rendre neutre.
    expect(() => stationCode("          ")).toThrow(/forme/i);
    expect(() => stationCode("undefined0")).toThrow(/forme/i);
    expect(() => stationCode("k447001001")).toThrow(/forme/i);
  });
});

File: test/domain/station/station_test.dart
// Verrouille la forme mesurée du code station (C-05) et la forme du code
// département en chaîne. C-05 reproduit le 2026-09-13 : le code site
// K4470010 (8 car.) renvoie chaque mesure en double, dont une ligne à
// code_station null — count 430 contre 216 pour le code station.
import 'package:flutter_test/flutter_test.dart';
import 'package:martinpecheur/domain/station/station.dart';

void main() {
  group('StationCode (C-05)', () {
    test("'K447001001' est accepté — forme la plus fréquente", () {
      expect(StationCode('K447001001').value, 'K447001001');
    });

    test("'1011000101' est accepté — dix chiffres, forme des DOM", () {
      expect(StationCode('1011000101').value, '1011000101');
    });

    test("'K4470010' (8 car.) est refusé — c'est le code site de C-05", () {
      expect(() => StationCode('K4470010'), throwsArgumentError);
    });

    test('une chaîne vide ou de onze caractères est refusée', () {
      expect(() => StationCode(''), throwsArgumentError);
      expect(() => StationCode('K4470010012'), throwsArgumentError);
    });

    test("dix caractères mais mauvaise forme ('k447-01001', 'k447001001') "
        'sont refusés', () {
      expect(() => StationCode('k447-01001'), throwsArgumentError);
      expect(() => StationCode('k447001001'), throwsArgumentError);
    });

    test('deux codes de même valeur sont égaux et de même hashCode', () {
      final StationCode a = StationCode('K447001001');
      final StationCode b = StationCode('K447001001');

      expect(a, b);
      expect(a.hashCode, b.hashCode);
    });
  });

  group('DepartementCode', () {
    test("'41' reste '41', '2a' devient '2A', '2B' reste '2B', "
        "'971' reste '971'", () {
      expect(DepartementCode('41').value, '41');
      expect(DepartementCode('2a').value, '2A');
      expect(DepartementCode('2B').value, '2B');
      expect(DepartementCode('971').value, '971');
    });

    test("'1' (un seul chiffre) et 'ZZ' sont refusés", () {
      expect(() => DepartementCode('1'), throwsArgumentError);
      expect(() => DepartementCode('ZZ'), throwsArgumentError);
    });
  });

  group('Station', () {
    test('porte les champs du référentiel — K447001001, relevés le '
        '2026-09-13', () {
      final Station station = Station(
        code: StationCode('K447001001'),
        label: 'Le Loir à Vaas',
        latitude: 47.584957074,
        longitude: 1.335147948,
        departement: DepartementCode('41'),
        riverLabel: 'Le Loir',
        inService: true,
      );

      expect(station.code, StationCode('K447001001'));
      expect(station.latitude, 47.584957074);
      expect(station.longitude, 1.335147948);
      expect(station.departement, DepartementCode('41'));
      expect(station.inService, true);
    });

    test('sans cours d\'eau, riverLabel est null — jamais une chaîne vide '
        '(BR-007)', () {
      final Station station = Station(
        code: StationCode('1011000101'),
        label: 'Station DOM',
        latitude: 16.189402,
        longitude: -61.658989,
        departement: DepartementCode('971'),
        riverLabel: null,
        inService: true,
      );

      expect(station.riverLabel, isNull);
    });
  });
}
